fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/LoanReadPlatformServiceImpl.java

```diff
@@ -2439,14 +2439,38 @@ public LoanTransactionData retrieveLoanForeclosureTemplate(final Long loanId, fi
         final Boolean isReversed = false;
 
         final Money outStandingAmount = loanRepaymentScheduleInstallment.getTotalOutstanding(currency);
+        BankAccountReadPlatformServiceImpl.BankAccountMapper bankAccountMapper = new BankAccountReadPlatformServiceImpl.BankAccountMapper();
 
-        return new LoanTransactionData(null, null, null, transactionType, null, currencyData, earliestUnpaidInstallmentDate,
-                outStandingAmount.getAmount(), loan.getNetDisbursalAmount(),
-                loanRepaymentScheduleInstallment.getPrincipalOutstanding(currency).getAmount(),
-                loanRepaymentScheduleInstallment.getInterestOutstanding(currency).getAmount(),
-                loanRepaymentScheduleInstallment.getFeeChargesOutstanding(currency).getAmount(),
-                loanRepaymentScheduleInstallment.getPenaltyChargesOutstanding(currency).getAmount(), null, unrecognizedIncomePortion,
-                paymentTypeOptions, null, null, null, outstandingLoanBalance, isReversed);
+        String hierarchy = this.context.authenticatedUser().getOffice().getHierarchy();
+        String bankAccSql = "select " + bankAccountMapper.schema();
+        bankAccSql = bankAccSql + " where (mo.hierarchy LIKE CONCAT(?, '%') OR ? like CONCAT(mo.hierarchy, '%')) ";
+        List<BankAccountData> bankAccounts = this.jdbcTemplate.query(bankAccSql, bankAccountMapper, hierarchy, hierarchy);
+
+        BigDecimal penaltyAmount = loanRepaymentScheduleInstallment.getPenaltyChargesOutstanding(currency).getAmount();
+        BigDecimal feeCharges = loanRepaymentScheduleInstallment.getFeeChargesOutstanding(currency).getAmount();
+        BigDecimal interestOutstanding = loanRepaymentScheduleInstallment.getInterestOutstanding(currency).getAmount();
+
+        // if percentage of paid installments is less than 50 add charge
+        if (loan.getLoanRepaymentScheduleDetail().getInterestMethod().equals(InterestMethod.DECLINING_BALANCE)) {
+            String completedPaymentsSql = "select count(*) from m_loan_repayment_schedule where loan_id = ? and completed_derived = true";
+            Integer numberOfPayments = this.jdbcTemplate.queryForObject(completedPaymentsSql, Integer.class, loanId);
+
+            List<LoanRepaymentScheduleInstallment> repaymentScheduleInstallments = loan.getRepaymentScheduleInstallments();
+
+            Double percentage = Double.valueOf((numberOfPayments /= repaymentScheduleInstallments.size()) * 100);
+
+            if (percentage.compareTo(50D) < 0) {
```

Comment:
@BrianMuhimbura, it would be good to move these values to global parameters or something that we can update without changing the code?